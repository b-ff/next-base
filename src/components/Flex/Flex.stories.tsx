import type { Meta, StoryObj } from "@storybook/react";

import { Flex } from "./Flex";

const meta: Meta<typeof Flex> = {
  component: Flex,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof Flex>;

export const Primary: Story = {
  args: {
    children: ["Element 1", "Element 2", "Element 3"],
  },
};
