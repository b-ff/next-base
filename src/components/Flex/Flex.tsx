import { FC, HTMLProps, ReactNode } from "react";

type GapSizes = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export type FlexProps = HTMLProps<HTMLElement> & {
  padding?: GapSizes;
  margin?: GapSizes;
  gap?: GapSizes;
  direction?: "row" | "row-reverse" | "column" | "column-reverse";
  children: ReactNode | ReactNode[];
};

export const Flex: FC<FlexProps> = (props) => {
  const {
    padding = 0,
    margin = 0,
    gap = 1,
    direction = "column",
    style,
    children = [],
    ...rest
  } = props;

  const combinedStyle = {
    ...style,
    display: "flex",
    flexDirection: direction,
    padding: `var(--spacing_${padding}x)`,
    margin: `var(--spacing_${margin}x)`,
    gap: `var(--spacing_${gap}x)`,
  };

  const elements =
    Array.isArray(children) && children.length ? children : [children];

  return (
    <section style={combinedStyle} {...rest}>
      {elements.map((element, idx) => (
        <div key={idx}>{element}</div>
      ))}
    </section>
  );
};
