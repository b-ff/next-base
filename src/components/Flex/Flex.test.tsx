import { Flex } from "./Flex";
import { render } from "@testing-library/react";

const mockChildren = ["Foo", "Bar", "Zoo", "Zar"];
const testId = "flex-test";

describe("Component Flex", () => {
  it("Should render as expected", () => {
    const { container } = render(<Flex>{mockChildren}</Flex>);
    expect(container).toMatchSnapshot();
  });

  it("Should render given children", () => {
    const dom = render(<Flex title={testId}>{mockChildren}</Flex>);
    const container = dom.getByTitle(testId);
    expect(container.childNodes.length).toBe(4);
  });
});
