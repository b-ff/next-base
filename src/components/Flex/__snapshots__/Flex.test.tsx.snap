// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Flex Should render as expected 1`] = `
<div>
  <section
    style="display: flex; flex-direction: column; gap: var(--spacing_1x);"
  >
    <div>
      Foo
    </div>
    <div>
      Bar
    </div>
    <div>
      Zoo
    </div>
    <div>
      Zar
    </div>
  </section>
</div>
`;
